### PR TITLE
Create a data access layer, which always return the user's filtered data for `/users` route

### PR DESCRIPTION
**Pull request by mailmesriza98**
coveragereport
related to https://github.com/Real-Dev-Squad/website-backend/issues/1196

modifications for /users route

### CODE REVIEW

File: services/dataAccessLayer.js [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      if (!result.userExists) {
+        return res.boom.notFound("User doesn't exist");
+      }
+
+      const user = result.user;
+      delete user.phone;
+      delete user.email;
+      delete user.chaincode;
+      delete user.tokens;
+
```
Comment (Pratiyushkumar) [inline services/dataAccessLayer.js:34]: i am not sure but if this can be done in one line then please do it

File: controllers/users.js [hunk 2]
```diff
@@ -5,13 +5,12 @@ const logsQuery = require("../models/logs");
 const imageService = require("../services/imageService");
 const { profileDiffStatus } = require("../constants/profileDiff");
 const { logType } = require("../constants/logs");
-
+const dataAccess = require("../services/dataAccessLayer");
 const logger = require("../utils/logger");
 const obfuscate = require("../utils/obfuscate");
-const { getPaginationLink, getUsernamesFromPRs } = require("../utils/users");
-const { getQualifiers } = require("../utils/helper");
+// const { getUsernamesFromPRs } = require("../utils/users");
```
Comment (Pratiyushkumar) [inline controllers/users.js:11]: remove commented code please

File: controllers/users.js [hunk 3]
```diff
@@ -5,13 +5,12 @@ const logsQuery = require("../models/logs");
...
 const { SOMETHING_WENT_WRONG, INTERNAL_SERVER_ERROR } = require("../constants/errorMessages");
-const { getFilteredPRsOrIssues } = require("../utils/pullRequests");
+// const { getFilteredPRsOrIssues } = require("../utils/pullRequests");
```
Comment (Pratiyushkumar) [inline controllers/users.js:13]: remove commented code please

File: services/dataAccessLayer.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        users,
+      });
+    }
+
+    const { allUsers, nextId, prevId } = await userQuery.fetchPaginatedUsers(req.query);
+    allUsers.forEach((element) => {
+      delete element.phone;
+      delete element.email;
+      delete element.chaincode;
+      delete element.tokens;
```
Comment (Pratiyushkumar) [inline services/dataAccessLayer.js:61]: if possible here also please

File: services/dataAccessLayer.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        prev: prevId ? getPaginationLink(req.query, "prev", prevId) : "",
+      },
+    });
+  } catch (error) {
+    logger.error(`Error while fetching all users: ${error}`);
+    return res.boom.serverUnavailable(SOMETHING_WENT_WRONG);
+  }
+};
+
+const removeSensitiveInfo = function (obj, props) {
```
Comment (bksh05) [inline services/dataAccessLayer.js:68]: A suggestion here

Can we have the function definition as given below?
This way, you don't have to pass an array every time you want to call the removeSensitiveInfo function, and it also gives you the flexibility to change parameters when required by passing an array as the second parameter.

```
function removeSensitiveInfo(data, properties = ["phone", "email", "chaincode", "tokens"]) {
  // Logic to remove sensitive information based on the properties array
  // ...
}
```

Example: 
```
 removeSensitiveInfo(element) // removes "phone", "email", "chaincode", and  "tokens"  as the default behavior
 removeSensitiveInfo(element, [token, chaincode])  // For cases where tokens and chaincode need to be removed.
```

**Review by bksh05 — COMMENTED** [review]
Please add tests for your code as well.

**Review by iamitprakash — CHANGES_REQUESTED** [review]
please add test coverage stats

File: services/dataAccessLayer.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        message: "User returned successfully!",
+        user,
+      });
+    }
+
+    if (qualifiers?.filterBy) {
+      const allPRs = await getFilteredPRsOrIssues(qualifiers);
+      const usernames = getUsernamesFromPRs(allPRs);
+      const { users } = await userQuery.fetchUsers(usernames);
+      removeSensitiveInfo(users, ["phone", "email", "chaincode", "tokens"]);
```
Comment (bksh05) [inline services/dataAccessLayer.js:42]: Since there is the same array in the function's default parameter, you don't have to pass this array. 
You can remove it here and in other instances too.

**Review by bksh05 — CHANGES_REQUESTED** [review]

File: services/dataAccessLayer.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        logger.error(`Error while fetching user: ${error}`);
+        return res.boom.serverUnavailable(SOMETHING_WENT_WRONG);
+      }
+
+      if (!result.userExists) {
+        return res.boom.notFound("User doesn't exist");
+      }
+
+      const user = result.user;
+      removeSensitiveInfo(user, ["phone", "email", "chaincode", "tokens"]);
```
Comment (bharati-21) [inline services/dataAccessLayer.js:30]: Kindly add the array to a constant and re-use it.

File: services/dataAccessLayer.js [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+  for (let i = 0; i < properties.length; i++) {
+    if (Object.prototype.hasOwnProperty.call(obj, properties[i])) {
+      delete obj[properties[i]];
+    }
+  }
+};
+
+module.exports = {
+  retrieveUsers,
+};
```
Comment (bharati-21) [inline services/dataAccessLayer.js:78]: Was the file formatted? If not, kindly run prettier command to format. The formatting seems off.

**Review by bharati-21 — CHANGES_REQUESTED** [review]
Kindly add test cases for the changes

File: services/dataAccessLayer.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
+const userQuery = require("../models/users");
+
+const retrieveUsers = async ({ id = null, usernames = null, req = null }) => {
```
Comment (iamitprakash) [inline services/dataAccessLayer.js:3]: this function takes three arguments, but only two of them are actually used. The req argument is never used, so it could be removed from the function signature.

File: services/dataAccessLayer.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+  } else {
+    const { allUsers, nextId, prevId } = await userQuery.fetchPaginatedUsers(req.query);
+    allUsers.forEach((element) => {
+      removeSensitiveInfo(element);
+    });
+    return { allUsers, nextId, prevId };
+  }
+};
+
+const removeSensitiveInfo = function (obj, properties = ["phone", "email", "chaincode", "tokens"]) {
```
Comment (iamitprakash) [inline services/dataAccessLayer.js:23]: properties[i] property exists in the obj object. However, the hasOwnProperty method only checks if the property is defined in the object's prototype chain. This means that the hasOwnProperty method will not return true if the property is defined directly in the obj object.

To fix this bug, you can use the in operator instead of the hasOwnProperty method. The in operator checks if the property exists in the object, regardless of whether it is defined in the object's prototype chain

**Review by iamitprakash — CHANGES_REQUESTED** [review]

File: services/dataAccessLayer.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    removeSensitiveInfo(result.user);
+    return result;
+  } else if (usernames) {
+    const { users } = await userQuery.fetchUsers(usernames);
+    users.forEach((element) => {
+      removeSensitiveInfo(element);
+    });
+    return users;
+  } else {
+    const { allUsers, nextId, prevId } = await userQuery.fetchPaginatedUsers(req.query);
```
Comment (bksh05) [inline services/dataAccessLayer.js:15]: You should check if req is null or not. 
Let's say someone calls your function ` retrieveUsers({})`, then `req.query` will throw an error here. 

Also since you are just using the query, then why not pass only the query instead of the entire request object?

**Review by bksh05 — CHANGES_REQUESTED** [review]

**Review by iamitprakash — DISMISSED** [review]

**Review by bksh05 — APPROVED** [review]
LGTM

**Review by bksh05 — APPROVED** [review]

**Review by bharati-21 — DISMISSED** [review]
LGTM

File: services/dataAccessLayer.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+  } else {
+    const { allUsers, nextId, prevId } = await userQuery.fetchPaginatedUsers(query);
+    allUsers.forEach((element) => {
+      removeSensitiveInfo(element);
+    });
+    return { allUsers, nextId, prevId };
+  }
+};
+
+const removeSensitiveInfo = function (obj, properties = ["phone", "email", "chaincode", "tokens"]) {
```
Comment (prakashchoudhary07) [inline services/dataAccessLayer.js:23]: Please move the properties to be removed to a constant

File: test/unit/services/dataAccessLayer.test.js [hunk 13]
```diff
@@ -0,0 +1,62 @@
+const chai = require("chai");
+const chaiHttp = require("chai-http");
+
+const userQuery = require("../../../models/users");
+const sinon = require("sinon");
+const { retrieveUsers, removeSensitiveInfo } = require("../../../services/dataAccessLayer");
+const userData = require("../../fixtures/user/user")();
+const remove = require("../../fixtures/user/removalData")();
```
[the remark this answers is not in the document]
Comment (prakashchoudhary07) [reply test/unit/services/dataAccessLayer.test.js:8]: `remove`

don't know why other lines are showing

File: constants/users.js [hunk 14]
```diff
@@ -16,4 +16,9 @@ const ALLOWED_FILTER_PARAMS = {
   ROLE: ["role"],
 };
 
-module.exports = { profileStatus, USER_STATUS, ALLOWED_FILTER_PARAMS };
+module.exports = {
+  profileStatus,
+  USER_STATUS,
+  ALLOWED_FILTER_PARAMS,
+  sensitiveData: ["phone", "email", "chaincode", "tokens"],
+};
```
Comment (Pratiyushkumar) [inline constants/users.js:24]: can we not make this like
`const sensitiveData: ["phone", "email", "chaincode", "tokens"],`
and then export `sensitiveData`

File: test/unit/services/dataAccessLayer.test.js [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+
+  describe("removeSensitiveInfo", function () {
+    it("should remove sensitive information from the users object", function () {
+      removeSensitiveInfo(userData);
+      sensitiveData.forEach((key) => {
+        expect(userData[12]).to.not.have.property(key);
+      });
+    });
+  });
+});
```
Comment (Pratiyushkumar) [inline test/unit/services/dataAccessLayer.test.js:62]: can we make this in separate test file, let's not crowd one test file with multiple test suites

File: test/unit/services/dataAccessLayer.test.js [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+const userData = require("../../fixtures/user/user")();
+const { sensitiveData } = require("../../../constants/users");
+chai.use(chaiHttp);
+const expect = chai.expect;
+
+describe("Data Access Layer", function () {
+  describe("retrieveUsers", function () {
+    it("should fetch a single user by ID and remove sensitive info", async function () {
+      const fetchUserStub = sinon.stub(userQuery, "fetchUser");
+      fetchUserStub.returns(Promise.resolve({ user: userData[12] }));
```
Comment (Pratiyushkumar) [inline test/unit/services/dataAccessLayer.test.js:16]: what will happen if multiple id's are passed

File: test/unit/services/dataAccessLayer.test.js [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+      removeSensitiveInfo(userData[12]);
+      expect(result.user).to.deep.equal(userData[12]);
+      sensitiveData.forEach((key) => {
+        expect(userData[12]).to.not.have.property(key);
+      });
+    });
+
+    it("should fetch users by username and remove sensitive info", async function () {
+      const fetchUserStub = sinon.stub(userQuery, "fetchUsers");
+      fetchUserStub.returns(Promise.resolve({ users: [userData[12]] }));
```
Comment (Pratiyushkumar) [inline test/unit/services/dataAccessLayer.test.js:28]: if we try to fetch by user id, won't it be good enough because id's are always unique or username should be made unique and use it

File: test/unit/services/dataAccessLayer.test.js
Comment (Pratiyushkumar) [inline test/unit/services/dataAccessLayer.test.js:1]: can we also test if pass retrieveUsers(id=null, userName=null, query=null) then what will happen

File: test/unit/services/dataAccessLayer.test.js (hunk 16 above)
Comment (mailmesriza98) [reply test/unit/services/dataAccessLayer.test.js:16]: multiple ids would not be passed in this case as per the function signature of retrieveUsers
